dipeptides/createDipeptides.py

```diff
@@ -0,0 +1,114 @@
+import openmm
+import openmm.app as app
+import openmm.unit as unit
+import numpy as np
+import h5py
+from io import StringIO
+
+def createDipeptide(forcefield, res1, res2, var1, var2):
```

Comment:
This is an interesting and weird way to make peptides. Why not construct from SMILES instead and have a great deal of flexibility? The OpenFF Toolkit `Molecule.from_smiles()` makes this easy, and it can generate conformations directly with `Molecule.generate_conformers`:

https://open-forcefield-toolkit.readthedocs.io/en/0.10.0/api/generated/openff.toolkit.topology.Molecule.html#openff.toolkit.topology.Molecule

It also handles the conversion to/from OpenMM automatically, and supports a bunch of file formats.

Since all the QCFractal/QCArchive tools we'll be using for this project use the OpenFF toolkit, may as well get some familiarity with using it.

Reply:
Weird in what way?  It's really simple to implement, and works well.  Having to build a table of SMILES strings for all the different variants of all amino acids would be a lot more complicated.

Reply:
It's highly inflexible, and uses a "roll-your own" approach to interconversion with RDKit (instead of a well-supported toolkit) that passes through a transient file-based intermediate. Changing to capped peptides or tripeptides is impossible. That's why it's weird.

Reply:
Especially when the next step will be feeding it to a toolkit that natively supports all these things.

Reply:
That's absurd.  Capping the peptides is trivial.  I just add them to the input PDB file.  Tripeptides are equally trivial.  And I find it kind of insulting that you just claimed that PDBFixer is not a "well-supported toolkit".  This is *exactly* the sort of problem that PDBFixer was designed to solve.  It makes it really really easy to do simple manipulations like this.

> Especially when the next step will be feeding it to a toolkit that natively supports all these things.

The next step is to feed it into OpenMM to build a System with CHARMM36.  PDBFixer gives us exactly the native format needed for that.

Reply:
PDBFixer isn't a well-supported toolkit for this kind of work---it's a part-time project that isn't very good at generating mutations (it literally rotates a reference amino acid rigidly into place), cannot generate arbitrary amino acids/caps, and cannot generate conformers. It doesn't have proper API documentation and the CI unit test haven't been run for over a year. It's not absurd to suggest it isn't the right tool for the job.

I know it "wasn't invented here", but at some point, we're going to have to move past that and embrace well-supported tools that are well-suited to the task at hand, which will be generating a diverse library of peptides---including capped peptides, cyclic peptides, disulfide peptides, post-translational modifications. It's best if we start off on the right foot with a tool that will be able to grow to that task.

> The next step is to feed it into OpenMM to build a System with CHARMM36. PDBFixer gives us exactly the native format needed for that.

While CHARMM36 is probably more reliable than OpenFF for parameterizing peptides, this will rely on having CHARMM36 parameters available for everything we want to build, which is going to break at some point. Is OpenFF a lot worse than CHARMM36 for this purpose? `openff-2.0.0` is out now.